acra-censor/common/logging_logic_test.go

```diff
@@ -0,0 +1,251 @@
+package common
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+
+func TestSerializationOnUniqueQueries(t *testing.T) {
+	testQueries := []string{
+		"SELECT Student_ID FROM STUDENT;",
+		"SELECT * FROM STUDENT;",
+		"SELECT * FROM X;",
+		"SELECT * FROM Y;",
+		"SELECT EMP_ID, NAME FROM EMPLOYEE_TBL WHERE EMP_ID = '0000';",
+		"SELECT EMP_ID, LAST_NAME FROM EMPLOYEE WHERE CITY = 'Seattle' ORDER BY EMP_ID;",
+		"SELECT EMP_ID, LAST_NAME FROM EMPLOYEE_TBL WHERE CITY = 'INDIANAPOLIS' ORDER BY EMP_ID asc;",
+		"SELECT Name, Age FROM Patients WHERE Age > 40 GROUP BY Age ORDER BY Name;",
+		"SELECT COUNT(CustomerID), Country FROM Customers GROUP BY Country;",
+		"SELECT SUM(Salary)FROM Employee WHERE Emp_Age < 30;",
+		"SELECT AVG(Price)FROM Products;",
+		"INSERT SalesStaff1 VALUES (2, 'Michael', 'Blythe'), (3, 'Linda', 'Mitchell'),(4, 'Jillian', 'Carson'), (5, 'Garrett', 'Vargas');",
+		"INSERT INTO SalesStaff2 (StaffGUID, FirstName, LastName) VALUES (NEWID(), 'Stephen', 'Jiang');",
+		"INSERT INTO SalesStaff3 (StaffID, FullName) VALUES (X, 'Y');",
+		"INSERT INTO SalesStaff3 (StaffID, FullNameTbl) VALUES (X, M);",
+		"INSERT INTO X.Customers (CustomerName, ContactName, Address, City, PostalCode, Country) VALUES ('Cardinal', 'Tom B. Erichsen', 'Skagen 21', 'Stavanger', '4006', 'Norway');",
+		"INSERT INTO Customers (CustomerName, City, Country) VALUES ('Cardinal', 'Stavanger', 'Norway');",
+		"INSERT INTO Production (Name, UnitMeasureCode,	ModifiedDate) VALUES ('Square Yards', 'Y2', GETDATE());",
+		"INSERT INTO T1 (Name, UnitMeasureCode,	ModifiedDate) VALUES ('Square Yards', 'Y2', GETDATE());",
+		"INSERT INTO dbo.Points (Type, PointValue) VALUES ('Point', '1,5');",
+		"INSERT INTO dbo.Points (PointValue) VALUES ('1,99');",
+	}
+	tmpFile, err := ioutil.TempFile("", "censor_log")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = tmpFile.Close(); err != nil {
+		t.Fatal(err)
+	}
+	writer, err := NewFileQueryWriter(tmpFile.Name())
+
+	defer func() {
+		writer.Release()
+		err = os.Remove(tmpFile.Name())
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, query := range testQueries {
+		_, err = writer.RedactAndCheckQuery(query)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	time.Sleep(DefaultSerializationTimeout + 100*time.Millisecond)
```

Comment:
in previous PR's we decreased time of tests twice just by decreasing granulation of timeouts. and there we can't it handle more deterministic because we can't control environment fully
but here in unit-tests we can. what about to move background goroutine in our constructor `NewFileQueryWriter` to some function like `start()` and rewrite like something:
```
func (writer *QueryWriter) start(){
  for {
			select {
			case <-writer.serializationTicker.C:
				err := writer.dumpBufferedQueries()
				if err != nil {
					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump buffered queries")
				}
				writer.serializationTicker.Stop()
				writer.serializationTicker = time.NewTicker(writer.serializationTimeout)
			case <-signalBackgroundExit:
				writer.serializationTicker.Stop()
				err := writer.logStorage.Close()
				if err != nil {
					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on exit QueryWriter instance")
				}
			case <-signalShutdown:
				writer.serializationTicker.Stop()
				err := writer.logStorage.Close()
				if err != nil {
					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on shutdown QueryWriter instance")
				}
				return
			default:
				//do nothing. This means that channel has no data to read yet
			}
		}
}

// NewFileQueryWriter creates QueryWriter instance
func NewFileQueryWriter(filePath string) (*QueryWriter, error) {
  writer := &QueryWriter{}
// initialization logic
  go writer.Start()
  return writer, nil
}
```

after that we can in our tests do:
```
writer, err := NewFileQueryWriter(tmpFile.Name())
for _, query := range testQueries {
		_, err = writer.RedactAndCheckQuery(query)
		if err != nil {
			t.Fatal(err)
		}
	}
writer.Release() // or writer.DumpQueries() or any other private method
go writer.start()  // here we restart our background goroutine 
```
plus we need this background goroutine only to test logic with background serialization and stop. in other tests (dumping/serialization/deserialization) we can test without it, just call direct methods with `writer.dumpBlaBla`, `writer.deserializeBlaBla`